### ISSUE TITLE
Hazard Class - Tiff (Transform = Affine)

### ISSUE THREAD

**Issue by lauradealmeidapereira** [CLIMADA-project/climada_python#806]
Dear Climada developers,

I am currently trying to create a Hazard from a Tiff file. However, when I plot it I can see that only the South Hemisphere appears. Could you please help me to find out what I am doing wrong or what I am missing?

```
haz_oc_fl = Hazard.from_raster(oc_fl_proc_file, haz_type = 'FL', attrs={'frequency':1}, transform=Affine(1, 0, -1, 0, -1, 0),
                            height=180, width=360)
haz_oc_fl.check()

haz_oc_fl.intensity[haz_oc_fl.intensity==-9999] = 0
haz_oc_fl.plot_intensity(1, smooth=False) 
```

![image](https://github.com/CLIMADA-project/climada_python/assets/72861957/006072cf-8139-4138-9dd4-c23d9557c027)

**Comment by chahank** [comment]
Hi! 
Without knowing what your file is, it is very hard to know what is happening. Have you double-checked that your input file is correct?

**Comment by lauradealmeidapereira** [comment]
Hi Chahan,

Thank you for your prompt response. 

I have created the mentioned tiff file with:
```
import xarray as xr
import rioxarray as rio
from affine import Affine
import os

ncfile = xr.open_dataset(HAZ_SEA_LEVEL_85_2055)

pr = ncfile['slr_he']

pr = pr.rio.set_spatial_dims('x', 'y')

#Check for the CRS
#pr.rio.crs

pr.rio.set_crs("epsg:4326")

dir_processed = '...\\hazard_data\\Coastal_flood'
oc_fl_proc_file = os.path.join(dir_processed,'rsl_85_2055.tif')
pr.rio.to_raster(oc_fl_proc_file)
```
The image looks like:
![image](https://github.com/CLIMADA-project/climada_python/assets/72861957/03bb7e51-e457-4292-82e2-b1cdaa7f6b26)

I have added both, the original nc file and the Tiff file in the zip file below. 

[rsl_85_2055.zip](https://github.com/CLIMADA-project/climada_python/files/13377797/rsl_85_2055.zip)

**Comment by chahank** [comment]
It looks like your coordinates are inverted. Maybe this leads to wrong reading of the hazard. I suggest the following:
- Can you please post the picture of `haz_oc_fl.centroids.plot()` ?
- Try to load your hazard directly from xarray instead? `Hazard.from_xarray_raster_file(HAZ_SEA_LEVEL_85_2055, intensity='slr_he')`

**Comment by lauradealmeidapereira** [comment]
Thank you for the tips.

The `haz_oc_fl.centroids.plot()` looks like:

![image](https://github.com/CLIMADA-project/climada_python/assets/72861957/524fbab0-c716-4ef9-82e7-9db33356b8a7)

And when I use `from_xarray_raster_file` as below

```
haz_oc_fl = Hazard.from_xarray_raster_file(HAZ_SEA_LEVEL_85_2055, intensity = 'slr_he', 
                                           hazard_type = 'FL',intensity_unit = 'm')

haz_oc_fl.check()
haz_oc_fl.intensity[haz_oc_fl.intensity==-9999] = 0
haz_oc_fl.plot_intensity(1, smooth=True)
```
I receive the following error:

> ---------------------------------------------------------------------------
> KeyError                                  Traceback (most recent call last)
> ~\AppData\Local\anaconda3\envs\climada_env\lib\site-packages\xarray\core\dataset.py in ?(self, name)
>    1432             variable = self._variables[name]
>    1433         except KeyError:
> -> 1434             _, name, variable = _get_virtual_variable(self._variables, name, self.dims)
>    1435 
> 
> KeyError: 'time'
> 
> During handling of the above exception, another exception occurred:
> 
> KeyError                                  Traceback (most recent call last)
> c:\users\de124956\documents\climada\climada_python\climada\hazard\base.py in ?(cls, data, hazard_type, intensity_unit, intensity, coordinate_vars, data_vars, crs, rechunk)
>     725         except KeyError as err:
>     726             key = err.args[0]
> --> 727             raise RuntimeError(
>     728                 f"Dataset is missing dimension/coordinate: {key}. Dataset dimensions: "
> 
> ~\AppData\Local\anaconda3\envs\climada_env\lib\site-packages\xarray\core\dataset.py in ?(self, key)
>    1528             return self.isel(**key)
>    1529         if utils.hashable(key):
> -> 1530             return self._construct_dataarray(key)
>    1531         if utils.iterable_of_hashable(key):
> 
> ~\AppData\Local\anaconda3\envs\climada_env\lib\site-packages\xarray\core\dataset.py in ?(self, name)
>    1432             variable = self._variables[name]
>    1433         except KeyError:
> -> 1434             _, name, variable = _get_virtual_variable(self._variables, name, self.dims)
>    1435 
> 
> ~\AppData\Local\anaconda3\envs\climada_env\lib\site-packages\xarray\core\dataset.py in ?(variables, key, dim_sizes)
>     204     split_key = key.split(".", 1)
>     205     if len(split_key) != 2:
> --> 206         raise KeyError(key)
>     207 
> 
> KeyError: 'time'
> 
> The above exception was the direct cause of the following exception:
> 
> RuntimeError                              Traceback (most recent call last)
> Cell In[13], line 1
> ----> 1 haz_oc_fl = Hazard.from_xarray_raster_file(HAZ_SEA_LEVEL_85_2055, intensity = 'slr_he', 
>       2                                            hazard_type = 'FL',intensity_unit = 'm')
>       4 haz_oc_fl.check()
>       6 # The masked values of the raster are set to 0
>       7 # Sometimes the raster file does not contain all the information, as in this case the mask value -9999
>       8 # We mask it manuall and plot it using plot_intensity()
> 
> File c:\users\de124956\documents\climada\climada_python\climada\hazard\base.py:449, in Hazard.from_xarray_raster_file(cls, filepath, *args, **kwargs)
>     412 """Read raster-like data from a file that can be loaded with xarray
>     413 
>     414 This wraps :py:meth:`~Hazard.from_xarray_raster` by first opening the target file
>    (...)
>     446 ...     hazard = Hazard.from_xarray_raster(dset, "", "")
>     447 """
>     448 with xr.open_dataset(filepath, chunks="auto") as dset:
> --> 449     return cls.from_xarray_raster(dset, *args, **kwargs)
> 
> File c:\users\de124956\documents\climada\climada_python\climada\hazard\base.py:727, in Hazard.from_xarray_raster(cls, data, hazard_type, intensity_unit, intensity, coordinate_vars, data_vars, crs, rechunk)
>     725 except KeyError as err:
>     726     key = err.args[0]
> --> 727     raise RuntimeError(
>     728         f"Dataset is missing dimension/coordinate: {key}. Dataset dimensions: "
>     729         f"{list(data.dims.keys())}"
>     730     ) from err
>     732 # Try promoting single-value coordinates to dimensions
>     733 for key, val in dims.items():
> 
> RuntimeError: Dataset is missing dimension/coordinate: time. Dataset dimensions: ['x', 'y']

**Comment by peanutfun** [comment]
@lauradealmeidapereira I think the original issue was your custom `transform`. Without it, everything looks fine to me:

```python
from climada.hazard import Hazard
hazard = Hazard.from_raster("rsl_85_2055.tif", band=0, haz_type="FL")
hazard.plot_intensity(event=0)
```
![output](https://github.com/CLIMADA-project/climada_python/assets/34276446/0f39e4ba-ca12-4093-90fe-1e6fedd2fb22)
Note that the transform usually is stored in the GeoTIFF file and will be read by the appropriate backend.

Alternatively, I also found a way to load the GeoTIFF with `from_xarray_raster`. This is a bit more complicated because this method requires a 3D dataset:

```python
import xarray as xr
from climada.hazard import Hazard
with xr.open_dataset("rsl_85_2055.tif", chunks="auto", engine="rasterio") as ds:
    ds = ds.squeeze()  # Remove 'band' dimension
    ds = ds.expand_dims(time=[np.datetime64("2055-01-01")])  # Add 'time' dimension
    hazard = Hazard.from_xarray_raster(
        ds,
        "FL",
        "",
        intensity="band_data",
        coordinate_vars=dict(longitude="x", latitude="y", event="time"),
    )

hazard.plot_intensity(event=0)
```
![output1](https://github.com/CLIMADA-project/climada_python/assets/34276446/fd4b5fbc-ba4c-411a-a458-51b3e204d92e)

**Comment by lauradealmeidapereira** [comment]
Thank you so very much, @peanutfun Lukas. 

Applying the first option
```
from climada.hazard import Hazard
hazard = Hazard.from_raster("rsl_85_2055.tif", band=0, haz_type="FL")
hazard.plot_intensity(event=0)
```
I get the error below when checking it (`hazard.check()`)

> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> Cell In[45], line 4
>       2 hazard = Hazard.from_raster(oc_fl_proc_file, band=0, haz_type="FL")
>       3 hazard.plot_intensity(event=0)
> ----> 4 hazard.check()
> 
> File c:\users\de124956\documents\climada\climada_python\climada\hazard\base.py:283, in Hazard.check(self)
>     276 def check(self):
>     277     """Check dimension of attributes.
>     278 
>     279     Raises
>     280     ------
>     281     ValueError
>     282     """
> --> 283     self.centroids.check()
>     284     self._check_events()
> 
> File c:\users\de124956\documents\climada\climada_python\climada\hazard\centroids\centr.py:177, in Centroids.check(self)
>     174     raise ValueError('Affine transformations with shearing components are not '
>     175                      'supported.')
>     176 if yres > 0 or xres < 0:
> --> 177     raise ValueError('Affine transformations with positive y-orientation '
>     178                      'or negative x-orientation are not supported.')
> 
> ValueError: Affine transformations with positive y-orientation or negative x-orientation are not supported.

Would you know how to correct this error?

Applying the second option it worked smoothly!

Thank you.

Best regards, 
Laura

**Comment by peanutfun** [comment]
@lauradealmeidapereira Thanks for reporting back. The check that throws the error might be deprecated.

@chahank Any reason why "Affine transformations with positive y-orientation or negative x-orientation are not supported."? To what kind of issues could this lead later on?

**Comment by chahank** [comment]
No, I do not know. Maybe @tovogt might know the answer?

**Comment by tovogt** [comment]
> @chahank Any reason why "Affine transformations with positive y-orientation or negative x-orientation are not supported."? To what kind of issues could this lead later on?

There is an open "question" issue about this: https://github.com/CLIMADA-project/climada_python/issues/100

**Comment by peanutfun** [comment]
So this is indeed some kind of a legacy issue, which also could not be resolved completely in #100. Since the check is related to `Centroids.meta`, which will be removed by #787, I think we can leave this as is for now.

**Comment by lauradealmeidapereira** [comment]
Thank you all so very much for your attention to this matter. 

Kind regards
Laura
